Proposal: Horizontal Pod Autoscaler Version 2

DirectXMan12 opened the pull request:
**Release note**:

``` release-note
NONE
```

This proposal introduces a redesigned Horizontal Pod Autoscaler,
designed to better allow scaling based on multiple metrics, and scaling
based on arbitrary metrics.

<!-- Reviewable:start -->

---

This change is [<img src="https://reviewable.kubernetes.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes/34754)

<!-- Reviewable:end -->

DirectXMan12 committed:
Proposal: Horizontal Pod Autoscaler Version 2

This proposal introduces a redesigned Horizontal Pod Autoscaler,
designed to better allow scaling based on multiple metrics, and scaling
based on arbitrary metrics.

DirectXMan12 commented:
cc @kubernetes/autoscaling as discussed at today's meeting, here's the updated version with the changes we agreed upon.

DirectXMan12 commented:
cc @kubernetes/kube-api @smarterclayton for API review

DirectXMan12 commented:
(hopefully that's the right team -- the features repo says `@kubernetes/api`, but that team doesn't exist)

balajismaniam commented:
xref PR: https://github.com/kubernetes/kubernetes/pull/34586.

thockin commented on docs/proposals/hpa-v2.md:
```diff
@@ -0,0 +1,228 @@
+doesn't provide a method for scaling based on memory percentage.
+
+Two alternatives to the method proposed above are shown the object below.
+The former is more compact, but precludes scaling memory as a percentage
```
I don't understand why memory is precluded...

bgrant0607 replied:
I think they mean they'd have to introduce another special field to specify a memory target.

DirectXMan12 replied:
it could be, just not from a "percentage of resource" style (it would be in as literal value).  For example, you could refer to memory like this regardless of how we do CPU:
```yaml
- type: pods
  pods:
    - metricName: memory/usage
      targetValue: 256 Mi
```.

If we had a `resourceRequest` type, we could *also* have:
```
- type: resourceRequest
  resourceRequest:
    name: memory
    targetPercentage: 50
````

which says "the target value is 50% of whatever the requested memory for the pod is", just like we do CPU.

thockin replied:
Ah, right.  Yeah that's a mess.  Percentage of request is useful.  What if there were two target fields - one for absoluteTarget and one for relativeTarget ?

DirectXMan12 replied:
(also, even with that, nothing prevents you from doing
```
- type: pods
  pods:
    metricName: memory/usage
    targetValue: 256 Mi
```
but there's not much reason to)

thockin replied:
something like that seems in the right direction to me.  We can wait for @fgrzadkowski to weigh in.

fgrzadkowski replied:
I like having CPU separately for several reasons. First of all CPU data comes from a separate pipeline (see new monitoring architecture for monitoring) that is supposed to be:
a) fast
b) always available (it's part of core deployment)
c) work in identical way everywhere
d) report information which are consistent with information in all other places that use them (e.g. scheduler, or kubelet, or user when specifying request/limit)

For custom metrics we don't have any of those guarantees. What's more they are far less portable (they work differently with different monitoring solutions).

For those reasons I believe we should treat CPU separately. Whether it's a special field in `Metric` or a top level field is less important for me, though IMO it's more clear if we have a group of metrics which are different in some way to be grouped together (like you have in this proposal).

With regard to memory it's actually very tricky to scale based on this, as it's very hard to build a proper model for scaling based on this signal. We've been doing some research around that and decided in the end it's not worth the effort. But I'm open for suggestions here :)

DirectXMan12 replied:
> For those reasons I believe we should treat CPU separately. Whether it's a special field in Metric or a top level field is less important for me, though IMO it's more clear if we have a group of metrics which are different in some way to be grouped together (like you have in this proposal).

If you're ok with having it be a separate source in Metrics, I'd prefer that (I agree with @thockin here -- it makes the API look nicer).  To clarify, are you ok with a general "resource metrics" source (so you can say `resource: cpu` or `resource: memory` or any other resource available from the master metrics API)?

balajismaniam commented on docs/proposals/hpa-v2.md:
```diff
@@ -0,0 +1,228 @@
+    Object *ObjectMetricSource `json:"object,omitempty"`
+
+    // metric describing pods in the scale target
+    Pods *PodsMetricSource `json:"pods,omitemtpy"`
```
Is there a reason why pods are considered separately from other objects?

thockin replied:
It's not clear to me how some other object reports metrics.  what is the concrete case for this and how does it work?

balajismaniam commented on docs/proposals/hpa-v2.md:
```diff
@@ -0,0 +1,228 @@
+    TargetCPUUtilizationPercentage `json:"targetCPUUtilizationPercentage,omitemtpy"`
+
+    // the non-CPU percentage metrics (optional)
+    Metrics []MetricSpec `json:"metrics,omitempty"`
```
Use `CustomMetrics` instead of `Metrics`?

DirectXMan12 replied:
Not necessarily custom (could be built-in metrics like memory, network, etc).

balajismaniam commented on docs/proposals/hpa-v2.md:
```diff
@@ -0,0 +1,228 @@
+    Object *ObjectMetricStatus `json:"object,omitemtpy"`
+
+    // metric describing pods in the scale target
+    Pods *PodsMetricStatus `json:"pods,omitemtpy"`
```
Same as above. Does pod status need to be separate from other object status?

DirectXMan12 replied:
(see above -- "pods" == "the group of pods referred to by the scale target")

balajismaniam commented:
cc @andrzej-k @IzabellaRaulin.

bgrant0607 commented on docs/proposals/hpa-v2.md:
```diff
@@ -0,0 +1,228 @@
+    MetricName string `json:"metricName"`
+
+    // the target value of the metric (as a quantity)
+    TargetValue resource.Quantity `json:"targetValue"`
```
Are such metrics expected to vary proportionally with load? What about request latency and error rate?

DirectXMan12 replied:
Perhaps I'm misunderstanding you here, but I don't think it should (or can be) our responsibility to decide which metrics a user can scale on.  We (as Kubernetes) authors can't really know which metrics might vary proportionally to load for a given application.  As a somewhat "silly" example, error rate might vary proportionally to load if the bulk of the errors are "max connections queued in server" (but you probably shouldn't autoscale on error rate, even then).

DirectXMan12 commented on docs/proposals/hpa-v2.md:
```diff
@@ -0,0 +1,228 @@
+type PodsMetricSource struct {
+
+    // the name of the metric in question
+    MetricName `json:"metricName"`
```
it's not supposed to be -- that was a typo ;-)

DirectXMan12 replied:
(it should read `MetricName string ...`)

thockin replied:
Fix and repush with type

DirectXMan12 commented on docs/proposals/hpa-v2.md:
```diff
@@ -0,0 +1,228 @@
+    // the target CPU percentage (optional)
+    TargetCPUUtilizationPercentage `json:"targetCPUUtilizationPercentage,omitemtpy"`
+
+    // the non-CPU percentage metrics (optional)
```
Maximum calculated desired replicas, as is the current functionality.  Previous versions include a separate field for specifying how the final number was calculated (max, min, average), but we removed it due to thinking it wasn't super-necessary.  We could revisit later.

thockin replied:
I meant the comments should reflect the semantic :)

DirectXMan12 commented on docs/proposals/hpa-v2.md:
```diff
@@ -0,0 +1,228 @@
+
+### Alternatives and Future Considerations ###
+
+Since the new design mirrors volume plugins (and similar APIs), it makes
```
that seems fairly reasonable.  It makes typing things in the YAML a bit redundant-looking, but the explicit tag is nice from a programming perspective.

thockin replied:
and it's cleaner across languages

thockin commented on docs/proposals/hpa-v2.md:
```diff
@@ -0,0 +1,228 @@
+    MaxReplicas int32 `json:"maxReplicas"`
+
+    // the target CPU percentage (optional)
+    TargetCPUUtilizationPercentage `json:"targetCPUUtilizationPercentage,omitemtpy"`
```
Filip,

The inconsistency really annoys me - is it really better?

DirectXMan12 replied:
I think the gist of the original argument was that since CPU is the common case, it should be "easy" to set up, and in the old proposal the new syntax was insufficiently "easy" to use for that case (and my effort at unifying in the original proposal yielded some ambiguities). I think the volume-style in this proposal makes it possible to just have a "metrics" field and still have an "easy-to-use syntax" (using either of the two alternatives I proposed).  Thoughts, @fgrzadkowski ?

thockin commented:
Overall this is OK.  @mwielgus or @jszczepkowski or @fgrzadkowski should have final OK on it, I'm reviewing mostly from an API clarity POV.

DirectXMan12 commented:
updated with @thockin's comments.  I think the last outstanding question is about CPU request (https://github.com/kubernetes/kubernetes/pull/34754#discussion_r84007627).

fgrzadkowski commented on docs/proposals/hpa-v2.md:
```diff
@@ -0,0 +1,242 @@
+    // the target CPU percentage
+    // +optional
+    TargetCPUUtilizationPercentage `json:"targetCPUUtilizationPercentage,omitemtpy"`
+    // the non-CPU percentage metrics (the maximum desired replica count
```
I think comment in the brackets should go somewhere before CPU, such that it's clear for the reader that it applies to all metrics including CPU.

DirectXMan12 replied:
ack

fgrzadkowski commented on docs/proposals/hpa-v2.md:
```diff
@@ -0,0 +1,242 @@
+    // the non-CPU percentage metrics (the maximum desired replica count
+    // across all metrics is used as desired replica count for the HPA)
+    // +optional
+    Metrics []MetricSpec `json:"metrics,omitempty"`
```
I'm wondering whether it would not be better if we name this field `CustomMetric`. I think it would give a clear message, that we have some "main metrics" and some "custom metrics".

fgrzadkowski commented on docs/proposals/hpa-v2.md:
```diff
@@ -0,0 +1,242 @@
+// (only `type` and one other matching field should be set at once)
+type MetricSpec struct {
+    // the type of metric source (should match one of the fields below)
+    Type MetricSourceType `json:"type"`
```
Why do we need this `Type` to be set? Wouldn't it be enough to just set `Pods` or `Object`?

fgrzadkowski replied:
ack

fgrzadkowski commented on docs/proposals/hpa-v2.md:
```diff
@@ -0,0 +1,242 @@
+    // +optional
+    TargetCPUUtilizationPercentage `json:"targetCPUUtilizationPercentage,omitemtpy"`
+    // the non-CPU percentage metrics (the maximum desired replica count
+    // across all metrics is used as desired replica count for the HPA)
```
I think we should explicitly say here what expectations do we have from such metrics, e.g. that it will drop with higher number of replicas.

DirectXMan12 replied:
ack

fgrzadkowski commented on docs/proposals/hpa-v2.md:
```diff
@@ -0,0 +1,242 @@
+it relatively easy to introduce new fields in a backwards-compatible way:
+we simply introduce a new field in `MetricSpec` as a new "metric type".
+
+#### External ####
```
I'd suggest deleting this section. I find it highly controversial and number of details are unclear (e.g. if we add `External` does it completely disable "regular" HPA controller?)

DirectXMan12 replied:
I'd like to record that we considered it and rejected it for the moment.  I'll make it clear that it has unresolved questions and that we decided it was not necessary for the moment.

fgrzadkowski commented on docs/proposals/hpa-v2.md:
```diff
@@ -0,0 +1,242 @@
+sufficient to allow specificying a `kind` of "Namespace" without a name.
+Alternatively, a similar source type to `PodsMetricSource` could be used.
+
+#### Aggregating Desired Replica Counts ####
```
I think it would extremely confusing for user to reason about the actions of HPA if we add custom aggregators between metrics.

fgrzadkowski commented on docs/proposals/hpa-v2.md:
```diff
@@ -0,0 +1,242 @@
+request percentage) will come from the
+[master metrics API](resource-metrics-api.md), while other metrics will
+come from the [custom/arbitrary metrics API](custom-metrics-api.md)
+(currently proposed as #34586).
```
... which will be a private adapter to the monitoring solution being currently used in the cluster.

DirectXMan12 replied:
✔️

fgrzadkowski commented on docs/proposals/hpa-v2.md:
```diff
@@ -0,0 +1,242 @@
+Horizontal Pod Autoscaler with Arbitary Metrics
```
Can we keep the name `Custom Metrics`? I think we've been using it for quite some already and adding new term `Arbitrary Metric` might be confusing.

DirectXMan12 replied:
I can.  The reason I went with arbitrary metrics is because with most monitoring systems, it will also be possible to access built-in metrics, so, for instance, if a user wanted to scale on one of those directly, not as a resource percentage, this could be used).

fgrzadkowski commented on docs/proposals/hpa-v2.md:
````diff
@@ -0,0 +1,242 @@
+
+```go
+type HorizontalPodAutoscalerSpec struct {
+    // the target scalable object to autoscale
````
Please change all of the field/struct comments to be regular sentences (e.g. start with capital letter and end with `.`). It will be more consistent with all other types definition. I know it's not the case currently, but we can fix it at the same time :)

DirectXMan12 replied:
that's not how most of the API objects, are documented, since they're exposed via the Swagger doc, IIRC.  Has that changed?

fgrzadkowski replied:
Oh. I only checked https://github.com/kubernetes/kubernetes/blob/master/pkg/api/types.go and it uses regular sentences.

DirectXMan12 commented:
updated to have resource metric as a distinct metric source, addressed some of @fgrzadkowski's suggested.

@fgrzadkowski most of the alternatives are there to show that we considered them but decided not to include them.  I've tried to make the language of a couple of those reflect that.

DirectXMan12 commented:
@fgrzadkowski the current revision look ok to you (especially re https://github.com/kubernetes/kubernetes/pull/34754#discussion_r85549992)?

derekwaynecarr commented:
@fgrzadkowski @thockin -- any other comments on this?

derekwaynecarr commented:
@DirectXMan12 -- fix the munger error

fgrzadkowski commented:
I can't some of the comments I see in my inbox. Grrr. I really hate github review experience!

Anyway.

1. This API is inconsistent with current API. Are you suggesting adding v2 api?
2. With regard to scaling based on memory, I'm still not convinced. I think we need a very clear line between metrics we fully support and guarantee they will work (cpu) vs other metrics (e.g. custom metrics). I think that memory fall rather into the latter category. It might work in some cases, but in most it will not. I don't want to set wrong expectations here. We should really not treat those sources equally. I know it looks nicer from the API perspective, but I think it'd set wrong expectations. IMO we should be very clear that CPU is something we really support. And custom metrics are "less supported" because we don't know how the metric will behave (e.g. we don't know if it's a linear function of number of replicas).
3. This proposal should discuss UX in `kubectl` for autoscaling related commands.

derekwaynecarr commented:
The corresponding PR proposed v2alpha1 API

andrzej-k commented on docs/proposals/hpa-v2.md:
```diff
@@ -0,0 +1,294 @@
+    name: WebFrontend
+  minReplicas: 2
+  maxReplicas: 10
+  metrics:
```
Shouldn't ```metrics``` also include ```type``` (see definition of ```MetricSpec``` type)?

DirectXMan12 replied:
yep, good catch.

andrzej-k commented on docs/proposals/hpa-v2.md:
```diff
@@ -0,0 +1,294 @@
+    // the name of the metric in question
+    MetricName string `json:"metricName"`
+    // the current value of the metric (as a quantity)
+    CurrentValue resource.Quantity `json:"targetValue"`
```
Maybe I'm missing something, but shouldn't `json:"targetValue"` be `json:"currentValue"` ?

DirectXMan12 replied:
yep, good catch

thockin replied:
still open

derekwaynecarr commented:
If we had to support non linear scaling functions in the future, is there a way we can change this to remove your skepticism?

fgrzadkowski commented:
@derekwaynecarr I don't think it matters what we can do in the future. This proposal IMO sets wrong expectations for the user. We are basically saying that resource metrics are always available etc so you can safely use them in all use-cases. As a user I'd probably set autoscaling for both cpu and memory so that I don't exceed requests for neither of them. In most cases this will not work as memory does not depend directly on the load and adding more replicas will not work. In many such scenarios we will just scale up to max (as memory will keep on the same high level) or we will do nothing (as it will be always below the threshold).

At the same time, if we just add custom metrics we will be able to use memory that would come from monitoring pipeline, so more advanced users who know how to use it will be able to workaround this limitation.

DirectXMan12 commented:
ping @thockin @fgrzadkowski can we figure something out here to move forward?

fgrzadkowski commented:
In order to address my concerns (assuming they are legitimate) I see two options:

1. Keep only two options - CPU (fully supported etc) and custom metrics (dependent on the monitoring pipeline); users who want to scale based on memory would use custom metrics for that.
2. Keep the API structs as they are proposed, but somehow explicitly explain in the comments/documentation what is the risk with scaling based on any metric other than CPU (memory in particular); this will require users to carefully read API documentation, but to some extent it should mitigate the problem I see.

I asked @thockin to look again at this PR and let us know what he thinks.

fgrzadkowski commented:
Yet another option is to postpone the decision about adding memory. For now extend `v1` api and only add custom metrics. Users will be able to scale based on memory using this. When we see there is high demand we can introduce `v2` api and add memory if needed.

@thockin Your feedback would be really appreciated.

thockin commented on docs/proposals/hpa-v2.md:
```diff
@@ -0,0 +1,294 @@
+// resource metric describing pods in the scale target
+// (guaranteed to be available and have the same names across clusters)
+type ResourceMetricSource struct {
+    // the name of the resource in question
```
examples?  "Resource" means so many things

DirectXMan12 replied:
as in requests and limits -- I'll put that in there.

thockin commented on docs/proposals/hpa-v2.md:
```diff
@@ -0,0 +1,294 @@
+// metric describing pods in the scale target
+type PodsMetricSource struct {
+    // the name of the metric in question
+    MetricName string `json:"metricName"`
```
This is unclear to me - which pod is the metric from?  I guess you assume that the metric is an aggregate metric, but in that case, it's not clear if or how you ensure it has anything to do with the pods...  Is this just an arbitrary metric and target value?

DirectXMan12 replied:
I've updated the godoc a bit, but it's a metric describing each pod (so each pod in the list should have a value for the associated metrics), which then gets aggregated together by the HPA.  Presumably, someone would not store a completely unrelated metric as being associated with each pod in the monitoring system. :shrug:

thockin commented on docs/proposals/hpa-v2.md:
```diff
@@ -0,0 +1,294 @@
+}
+
+// resource metric describing pods in the scale target
+type ResourceMetricSource struct {
```
Status ?

thockin commented:
Apparently I can not reply to outdated comments.  Thanks GitHub for being dumb.  I have a bunch of line-comments I will post after this.


WRT CPU and Memory:  We need to help users get things right and try to prevent them from getting things wrong.  In this case, I think good documentation is necessary but not sufficient. We also need to make sure that we log WHY we chose the scale we chose.  Events are imperfect but that's what we have for now, I guess.  If a user configures memory to 50% and all of their pods are going over 50%, I want them to see "I scaled this set because average memory usage is > 50%".  Given all that I think it is OK, and should satisfy Filip's #2.

WRT Naming: a lot of the names in here sort of suck (sorry).  Names need to make it clear what is happening.  Consider more precise names?  "AverageUtilization" is probably better than "TargetPercentageOfRequest" - it indicates the fact that values are aggregate, not per-pod.  Need to document this really well and in every place it is a factor.  As a user, looking up a single field in the docs, I need to see "this is the 30-second average across all pods in the set" or whatever the rules are.

WRT Resource metrics being different from plain old metrics.  I still don't like it.  I will accept if you collectively say it really is better that way, but it feels hacky.  At the very least it needs WAY better docs about what is different and WHY.

DirectXMan12 commented:
> WRT CPU and Memory: We need to help users get things right and try to prevent them from getting things wrong. In this case, I think good documentation is necessary but not sufficient. We also need to make sure that we log WHY we chose the scale we chose. Events are imperfect but that's what we have for now, I guess. If a user configures memory to 50% and all of their pods are going over 50%, I want them to see "I scaled this set because average memory usage is > 50%". Given all that I think it is OK, and should satisfy Filip's #2.

Yep, we currently do actually have events that more or less state this, and I've tried to preserve that in the WIP PR (and/or improve it in some cases).

> WRT Naming: a lot of the names in here sort of suck (sorry).

No offense taken.  Naming is one of the hard problems ;-).

> Names need to make it clear what is happening. Consider more precise names? "AverageUtilization" is probably better than "TargetPercentageOfRequest"- it indicates the fact that values are aggregate, not per-pod. Need to document this really well and in every place it is a factor.

Sure, that's fine.  I was going for "how do we calculate this" -- to me, `AverageUtilization` doesn't tell me anything -- what does utilization mean?  In this case utilization is "percentage of the requested value", and we call it a target since it's not a threshold.  I'll try and rework the names a bit, though.

> As a user, looking up a single field in the docs, I need to see "this is the 30-second average across all pods in the set" or whatever the rules are.

ack, I'll update the doc strings to be a bit more descriptive.

> WRT Resource metrics being different from plain old metrics. I still don't like it. I will accept if you collectively say it really is better that way, but it feels hacky. At the very least it needs WAY better docs about what is different and WHY.

Alright, so, there are a couple problems with fully unifying resource metrics and pod-aggregate metrics into one "source type":

1. resource metrics have known, constrained names (e.g. api.ResourceCPU, api.ResourceMemory, etc) of a *distinct type* (the fact that Go lacks proper enums to enforce this is besides the point), whereas other per-pod metrics don't have known names (and thus use the general string type), and may vary between collection interfaces.  This also ties into the fact that resource metrics are retrieved from a different location.

3. resource metrics are calculated differently -- resource metrics have the option to be dealt with as a percentage of request, but for non-resource metrics, percentage of request is a non-nonsensical operation -- they don't have a corresponding request value.

If we unified, we'd have to get rid of both of these distinctions.  We *could* check to see if `metricName` is actually a resource name in the HPA (but then lose the descriptiveness of listing it as such in the API), and *could* treat `targetValue` as a percentage in such a case, but we then lose the ability to refer to different views of it (percentage of request, raw, and perhaps percentage of limit in the future), and lose the descriptiveness of the name.

DirectXMan12 commented on docs/proposals/hpa-v2.md:
```diff
@@ -0,0 +1,294 @@
+type MetricSourceType string
+var (
+    // a metric describing a kubernetes object
+    ObjectSourceType MetricSourceType = "object"
```
(assuming you're talking about the value here): doesn't it make more sense for this value to match the corresponding field name in the YAML/JSON?

DirectXMan12 commented:
I've tried to fix and clarify the godoc strings to be more descriptive.  I haven't changed any names yet, since I'm still thinking about the best way to describe all the relevant things.

DirectXMan12 commented:
@thockin those names look a bit better? I updated the names of the targets in the `pod` and `resource` types.  I've been trying to come up with a better name for the `resource` type, but most of the names seem verbose or awkward, and it's probably going to be the most common type, so it should be quick to read and type.  Currently, it will look like:

```yaml
metrics:
- type: resource
  resource:
    name: cpu
    targetAverageRequestUtilization: 80
`

thockin commented on docs/proposals/hpa-v2.md:
```diff
@@ -0,0 +1,318 @@
+<!-- BEGIN MUNGE: UNVERSIONED_WARNING -->
```
please remove this header - it is obsolete

thockin commented on docs/proposals/hpa-v2.md:
```diff
@@ -0,0 +1,318 @@
+
+<!-- END STRIP_FOR_RELEASE -->
+
+<!-- END MUNGE: UNVERSIONED_WARNING -->
```
...all the way to here

thockin commented on docs/proposals/hpa-v2.md:
```diff
@@ -0,0 +1,318 @@
+type MetricSourceType string
+var (
+    // a metric describing a kubernetes object (for example, hits-per-second on an Ingress object)
+    ObjectSourceType MetricSourceType = "object"
```
API constants start with with an UpperCaseLetter

thockin commented on docs/proposals/hpa-v2.md:
```diff
@@ -0,0 +1,318 @@
+    // the target value of the resource metric, represented as
+    // a percentage of the requested value of the resource on the pods.
+    // +optional
+    TargetAverageRequestUtilization *int32 `json:"targetAverageRequestUtilization,omitempty"`
```
I find this to be a mouthful, but OK. I would drop "request" but reasonable people can disagree on this one

thockin reviewed (commented):
I'm OK with this.  Back to Filip for final approval

fgrzadkowski commented on docs/proposals/hpa-v2.md:
```diff
@@ -0,0 +1,318 @@
+    // a metric describing each pod in the current scale target (for example, transactions-processed-per-second).
+    // The values will be averaged together before being compared to the target value
+    Pods *PodsMetricSource `json:"pods,omitemtpy"`
+    // a resource metric known to Kubernetes, as specified in requests and limits, describing each pod
```
There are some resources that could be exported via resource metric API which are not CPU and memory and does not have requests & limits. Currently it might be `ResourceStorage` and `ResourceNvidiaGPU` (we already have resource names for those). I imagine there might be more that we expose via metric API but they would not be used in requests and limits.

I'd suggest to just drop this part of the sentence:

```
// a resource metric known to Kubernetes describing each pod in the current scale target (e.g. CPU or memory).
```

fgrzadkowski commented on docs/proposals/hpa-v2.md:
```diff
@@ -0,0 +1,318 @@
+// being compared to the target.  Such metrics are built in to Kubernetes, and have special
+// scaling options on top of those available to normal per-pod metrics using the "pods" source.
+// Only one "target" type should be set.
+type ResourceMetricStatus struct {
```
Are we gonna report both utilization and raw value or only the one which was specified in the spec? I think it would be good to specify it.

fgrzadkowski commented on docs/proposals/hpa-v2.md:
```diff
@@ -0,0 +1,318 @@
+interaction with the existing autoscaler, it was decided to exclude this
+feature.  We may reconsider in the future.
+
+### Limit Percentages ###
```
I think you are missing one more `#`.

fgrzadkowski reviewed (commented):
In general this LGTM. Please address remaining comments from Tim and myself.

thockin commented:
I ack that some don't like it, but that is the established and documented
API convention. :)



On Mon, Nov 28, 2016 at 3:21 PM, Solly Ross <notifications@github.com>
wrote:

> @thockin <https://github.com/thockin> @fgrzadkowski
> <https://github.com/fgrzadkowski> comments should be resolved.
>
> @thockin <https://github.com/thockin> re the uppercase API constants,
> I've changed the proposal to start with uppercase letters (so it should be
> all set now), but the reason I originally had lowercase letters is because
> it gives symmetry between the field name (in YAML/JSON) and the type
> value -- compare
>
> type: object
> object:
> ...
>
> with
>
> type: Object
> object:
> ...
>
> IMO, the upper one seems more ergonomic.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/34754#issuecomment-263426877>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AFVgVGE_gktXBfvg1VmrWf7wta2Rz5rzks5rC2ICgaJpZM4KWTAi>
> .
>

DirectXMan12 commented:
> I ack that some don't like it, but that is the established and documented API convention. :)

understood, that's a fine answer :-).  Just figured I'd bring it up, since you could make the argument that the "tag" on enums is different from other API constants.

fgrzadkowski commented:
LGTM

@kubernetes/autoscaling @mwielgus @jszczepkowski

thockin commented:
Is proposals offlimits now?  @kubernetes/docs

thockin commented:
https://groups.google.com/d/msgid/kubernetes-dev/CAKCBhs77rL1GkgQc%3DW0cXOP%2B5ox-MfUatjuUBTbNQzpK9xtfcA%40mail.gmail.com?utm_medium=email&utm_source=footer

On Wed, Nov 30, 2016 at 5:28 PM, Tim Hockin <notifications@github.com>
wrote:

> Is proposals offlimits now? @kubernetes/docs
> <https://github.com/orgs/kubernetes/teams/docs>
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/34754#issuecomment-264051593>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AFVgVAimKWJdEbEc6G0ENz9l-QHoeGjdks5rDiLagaJpZM4KWTAi>
> .
>

DirectXMan12 commented:
re-opening as kubernetes/community#120
